### ISSUE TITLE
Refining MAE Model with Custom Unlabeled Dataset

### ISSUE THREAD

**Issue by AlexMaks02** [ml-jku/MIM-Refiner#14]
First of all, thank you for your excellent work.

I have a few questions regarding refining a Masked Autoencoder (MAE) model using a custom unlabeled dataset.

- Are there any potential issues I should be aware of when using a new dataset for refinement?
- Are there any important modifications or considerations I should keep in mind when doing this?
- Does the existing training setup require any specific dataset format or preprocessing steps that might differ from the original dataset used in the paper?

Any guidance would be greatly appreciated. Thank you for your time!

**Comment by BenediktAlkin** [comment]
Hi, thanks for your interest in our work!

There are a few things to consider:

- The refinement process uses instance discrimination, which is known to work best on relatively "class-balanced" datasets. While it is no problem if the dataset is unlabeled, they work best if the distribution of concepts is somewhat balanced. This has been analyzed in the DINOv2 (https://arxiv.org/abs/2304.07193) paper where simply scraping an image dataset from the web is worse than curating it with their retrieval preprocessing pipeline. 
- We use a topk nearest neighbor swap for the refinement process to select one of the 20 nearest neighbors for the alignment part of the contrastive loss. Its probably good to start with top1 NN instead of top20 NN and once you get some good results you can increase that number. Simply change the topk in the queue_kwargs of the stage3 training (e.g. [here](https://github.com/ml-jku/MIM-Refiner/blob/main/src/yamls/stage3/h14_mae.yaml#L57))
- the temperature is a crucial parameter that depends on the quality of clusters. This is one of the most important hyperparameters to tune ("nnclr_temp03_cls" means, temperature=0.3), simply rename the head of stage2 training to change the temperature (e.g. [here](https://github.com/ml-jku/MIM-Refiner/blob/main/src/yamls/stage2/h14_mae.yaml#L42))
- the existing work implements the imagenet dataset format, i.e., it expects a folder that contains a subfolder for each class and each subfolder contains images. You can either preprocess your dataset into that format (as you dont have labels, just use a single subfolder and put all images into that; i.e. assign all images to an "unlabeled" class) or implement your own dataset ([here](https://github.com/ml-jku/MIM-Refiner/blob/main/src/datasets/imagenet.py) is the imagenet implementation). When you implement your own dataset, make sure to implement the getitem_class function and return 0 as the codebase expects a class label to calculate metrics during training. If your dataset is unlabeled, these metrics are not informative, but otherwise you'd need to remove the metric calculation from the code.

**Comment by AlexMaks02** [comment]
Thank you for your response! I really appreciate your time and insights. This helps a lot. Thanks again for your great work!
